### PR TITLE
Resource Management related cleanup

### PR DESCRIPTION
**Pull request by tdcmeehan**
Rebase of #12801

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 1]
```diff
@@ -174,6 +171,8 @@ public SqlQueryManager(
 
         this.clusterSizeMonitor = requireNonNull(clusterSizeMonitor, "clusterSizeMonitor is null");
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: note this line get removed in https://github.com/prestodb/presto/pull/12801/commits/bfa2524bae27cf40a021f51922d70af05dca27ee :)

File: presto-main/src/main/java/com/facebook/presto/dispatcher/QueuedStatementResource.java [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+
+    private static URI getQueuedUri(QueryId queryId, String slug, long token, UriInfo uriInfo, String xForwardedProto)
+    {
+        return uriInfo.getBaseUriBuilder()
+                .scheme(getScheme(xForwardedProto, uriInfo))
+                .replacePath("/v1/statement/queued/")
+                .path(queryId.toString())
+                .path(String.valueOf(token))
+                .replaceQuery("")
+                .queryParam("slug", slug)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/dispatcher/QueuedStatementResource.java]: Just double check: now slug is provided as a parameter, while in the old PR, slug is part of the path? (See https://github.com/prestodb/presto/pull/12801/commits/b84661660cb76f9247f1a9b21a85a1b9062f8ce2#diff-ef10213de38c18dc49c5303d7eea77a4R234)

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 3]
```diff
@@ -210,7 +125,6 @@ public void stop()
     {
         queryTracker.stop();
         queryManagementExecutor.shutdownNow();
-        unboundedExecutorService.shutdownNow();
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: hmm? :)

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java [hunk 4]
```diff
@@ -294,14 +263,12 @@ public synchronized boolean isClearTransactionId()
         return clearTransactionId;
     }
 
-    public synchronized ListenableFuture<QueryResults> waitForResults(OptionalLong token, UriInfo uriInfo, String scheme, Duration wait, DataSize targetResultSize)
+    public synchronized ListenableFuture<QueryResults> waitForResults(long token, UriInfo uriInfo, String scheme, Duration wait, DataSize targetResultSize)
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java]: Didn't see "slugValid" related method in this commit? (see https://github.com/prestodb/presto/pull/12801/commits/b84661660cb76f9247f1a9b21a85a1b9062f8ce2#diff-d65ad4547dcbc09fc621ac2ceddce6cdR217)

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java [hunk 5]
```diff
@@ -570,9 +502,9 @@ private synchronized URI createNextResultsUri(String scheme, UriInfo uriInfo)
     {
         return uriInfo.getBaseUriBuilder()
                 .scheme(scheme)
-                .replacePath("/v1/statement")
+                .replacePath("/v1/statement/executing")
                 .path(queryId.toString())
-                .path(String.valueOf(resultId.incrementAndGet()))
+                .path(String.valueOf(nextToken))
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java]: Why change this to `nextToken`?  :)

File: presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchExecutor.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+
+public class DispatchExecutor
+{
+    private final Closer closer = Closer.create();
+
+    private final ListeningExecutorService executor;
+    private final BoundedExecutor boundedExecutor;
```
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchExecutor.java]: Is it possible to also export the number of active threads in `BoundedExecutor`? -- this has been an operation pain when we want to monitor when `BoundedExecutor` is full.

File: presto-main/src/main/java/com/facebook/presto/dispatcher/QueuedStatementResource.java (hunk 2 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/dispatcher/QueuedStatementResource.java]: Yes, I saw from our commit history we did this because our HTTP logging seems to exclude query params from the logs.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java (hunk 3 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: This is not just a rebase conflict, looks like all references to `unboundedExecutorService` (was: `queryExecutor`) should have been removed entirely.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java (hunk 4 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java]: This is because we cherry picked this earlier in #13110

File: presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java (hunk 5 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/Query.java]: This was supposed to happen in a later commit.  Fixed it.

**Comment by tdcmeehan** [comment]
> A lot of abstractions that makes the change a little bit hard to follow, but generally looks good. It looks like the queries map in the `ExecutingStatementResource` is not being cleaned. It should be fixed to avoid memory leaks.

This is a really great catch.  Thank you.

File: presto-main/src/main/java/com/facebook/presto/server/protocol/ExecutingStatementResource.java [hunk 7]
```diff
@@ -273,19 +232,42 @@ private static Response toResponse(Query query, QueryResults queryResults)
...
+            return Response.noContent().build();
+        }
+        catch (NoSuchElementException e) {
+            throw badRequest(NOT_FOUND, "Query not found");
         }
-        query.cancel();
-        return Response.noContent().build();
+    }
+
+    private static WebApplicationException badRequest(Status status, String message)
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/server/protocol/ExecutingStatementResource.java]: What I'll do is, because every method that calls this passes in `NOT_FOUND` for the status, is I'll hardcode that `NOT_FOUND` and rename to `notFound`.

File: presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchExecutor.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchExecutor.java]: `BoundedExecutor` does not expose the number of active threads, however the underlying executor itself does export itself as an MBean via `ThreadPoolExecutorMBean`

File: presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchManager.java [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+                });
+    }
+
+    public void cancelQuery(QueryId queryId)
+    {
+        queryTracker.tryGetQuery(queryId)
+                .ifPresent(DispatchQuery::cancel);
+    }
+
+    private static class DispatchQueryCreationFuture
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/dispatcher/DispatchManager.java]: Turns out it's a different semantic.  An attempt to cancel this future has no effect, whereas cancelling the returned future from `nonCancellationPropagating` may be cancelled (however this will not be propagated to the inner future).

**Comment by wenleix** [comment]
Merged! Thanks for the contribution! 😃 😃 😃
